File: leet-code/relativeSortArray.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function (arr1, arr2) {
  return arr1.sort((a, b) => {
    if (!arr2.includes(a) && !arr2.includes(b)) {
      return a - b
    }
    if (!arr2.includes(a) && arr2.includes(b)) {
      return 1
    }
    if (!arr2.includes(b) && arr2.includes(a)) {
      return -1
    }
    if (arr2.includes(a) && arr2.includes(b)) {
      return arr2.indexOf(a) - arr2.indexOf(b)
    }
  })
}

console.log(
  relativeSortArray([2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19], [2, 1, 4, 3, 9, 6])
)

File: leet-code/relativeSortArray.go
package main

import "math"

func relativeSortArray(arr1 []int, arr2 []int) []int {
	maxV := math.MinInt64
	res := make([]int, 0)
	for _, v := range arr1 {
		if v > maxV {
			maxV = v
		}
	}

	countMap := make([]int, maxV+1)
	for _, v := range arr1 {
		countMap[v]++
	}

	for _, v := range arr2 {
		for j := 0; j < countMap[v]; j++ {
			res = append(res, v)
		}
		countMap[v] = 0
	}

	for i := range countMap {
		if countMap[i] <= 0 {
			continue
		}
		for j := 0; j < countMap[i]; j++ {
			res = append(res, i)
		}
	}
	return res
}
